### ISSUE TITLE
Should `this.complete()` be changed?

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#21335]
Today, `this.complete()` is used in initializers to indicate that the compiler should ensure an object is fully initialized.

It is treated as a syntactic marker and is not evaluated as a method call, despite the presence of `()`.

```chapel
record r {
  proc complete() { writeln('COMPLETE'); }
  proc init() {
    writeln('START');
    this.complete();
  }
}

proc main() {
  var x = new r();
}
```

Just prints `START`.

It is not considered a reserved word. The spec has several mentions of it as a "method", see:

https://chapel-lang.org/docs/language/spec/classes.html#limitations-on-instance-usage-in-initializers

Which mentions the `complete` method several times. This is at odds with the implementation's current treatment.

## Users may still want to define a method called `complete()`

Consider the handle for a database transaction `DBTransaction`. For such a type a method called `complete()` would be very commonplace and appreciated.

What should happen in the first code snippet? _Should_ the user's `complete()` method be called?

## Option 1: Refine the status quo

We can adjust the initializer spec sections to:

- Remove any mention of `complete()` as a method call.
- Say that within the body of an initializer, `this.complete()` is treated as a syntactic marker and is not evaluated as a method call.

But what about if the user wants to call their own `complete()` after the object is initialized?

E.g., if we called `this.complete()` twice in a row in the above code, the second expression should probably evaluate as a call to the user's `complete` method.

```chapel
this.complete(); // -> syntactic marker for compiler
this.complete(); // -> prints 'COMPLETE'
```

So we'd have to adjust the spec to say something like:

- If `this.complete()` appears and the object is not fully initialized, it is considered a syntactic marker and is not evaluated as a method call. If `this.complete()` has already appeared, subsequent uses are evaluated as a call to a method named `complete()`.

This is starting to get a little bit complicated.

## Option 2: Introduce a special method called `complete()`

This would introduce a new special method called `complete()` that would have to be handled in [#19038](https://github.com/chapel-lang/chapel/issues/19038). That way the user is still able to have `complete` as a potential method name.

We would adjust the language so that `this.complete()` is always evaluated as a method call.

So when `this.complete()` is encountered for the first time on any path in an initializer, the compiler will take its normal steps to ensure the aggregate is fully initialized, and then it will issue a call to `complete()`.

The default implementation of `complete()` would do nothing.

The only downside here is that I am not sure what the difference between `.complete()` and `.postinit()` is supposed to be.

## Option 2a: Make `this.complete()` a special method that cannot be called directly

This is like option 2, but:

- The special method for `this.complete()` cannot be called outside of initializers
- Its default implementation does nothing
- Users cannot provide their own implementations of it

## Option 3: Use a reserved word

In my opinion, the problem is not the idea that `this.complete()` is a syntactic marker - the issue is that it looks like a method call. We could fix this by using new syntax.

However, it is not simple enough to just drop the `()`. Recall that the line `this.complete;` can be evaluated as a parenless method invocation if a parenless method named `complete` is in scope.

I think that if we want to continue to treat `this.complete()` as a syntactic marker, we should probably transform it into a reserved word. Unless we took further steps, this would prevent users from using it as a method name.

To me, this seems acceptable as it brings `.complete()` in line with `init`, `deinit`, `postinit`.

Since it is reserved, we could even go so far as to adjust the syntax so that it doesn't look like a parenful method call at all:

```chapel
proc init() {
  complete; // Perhaps just a single keyword?
}
```

It might be good to use a different keyword / and or avoid reserving a new one such as `complete` which could be useful for many types. In which case we could use an existing keyword such as `init`:

```chapel
proc init() {
  init;
}
```

**Comment by bradcray** [comment]
At first blush, option 3 is most appealing to me (followed by option 1 then option 2), where I'd probably reserve a word other than `complete` (seems too common to reserve) and may even try to leverage existing reserved words (*).  E.g.:

```chapel
this.init;
init this;
type init;
init type;
record.init;
init record;
class.init;
init class;
init new;
init public;
break init;
init continue;
init lifetime;
init local;
require init;
init = true;
when init;
```

(* = where I assume that we're going to be reserving `init`

> I am not sure what the difference between .complete() and .postinit() is supposed to be.

In a class hierarchy, postinit() is called after the entire scaffolding of child/parent class initializers is called (i.e., after all inits complete), whereas `this.complete` would occur as each initializer runs.

The main drawback I see to option 2 is that it means that if a user wants to use `complete` as a method name, it will automatically get called by the initializer

**Comment by dlongnecke-cray** [comment]
> The main drawback I see to option 2 is that it means that if a user wants to use complete as a method name, it will automatically get called by the initializer

I was imagining that we would add `complete()` as another special method. We would use whatever strategy we decide on to avoid collisions for special methods in #19038, whether that be `this.~complete~()` or `complete(this)` + `chpl_complete()`, etc.

**Comment by bradcray** [comment]
Sorry, I didn't get that from the OP write-up of option 2.  If we had an option 2a that was:

* there is a built-in special method called `this.~complete~`
* users can't define their own overloads of it
* users can't invoke it anywhere but within their `init()` routines
* it serves as a syntactic marker, telling the compiler to complete the default initialization of any remaining fields and to consider the object initialized, permitting any subsequent statements to do anything they want with the object

I'd be about equally in favor of that as option 3.  I think the main downside to 2a is that it's pretty different than other special methods we're considering.  Then again, most of the special methods we're considering are different from one another in various ways, so maybe that's not the worst thing in the world...

**Comment by dlongnecke-cray** [comment]
I'm a big fan of the `this.init;` idea or maybe just `init;`. It's using an already reserved keyword, the compiler can print out a helpful message in the event that a user ever tries to do something like:

```chapel
proc init { ... } // A parenless method called this.init?!
```

And we sidestep any concerns about potentially stealing a method name from the user. The only downside is the footprint that will have to change, but that seems largely unavoidable.

It should also be relatively easy to have the compiler suggest you migrate from `this.complete()` to `this.init;`, as it pretty uniformly treats `this.complete()` as a syntactic marker only when in an initializer body.

The only downside I can think of is the potential confusion between `this.init` and `this.init()`. That makes me think that perhaps we could have `this.finish` or `this.finalize`, but that introduces another keyword...

**Comment by bradcray** [comment]
> The only downside I can think of is the potential confusion between `this.init` and `this.init()`

Oh, that's a good point.  I was thinking that users couldn't ever call this.init(), but forgot about the case of calling a sibling initializer...  `init this;` doesn't have that problem though!  :D

**Comment by lydia-duncan** [comment]
I don't think we should change it.  I don't actually think a user is going to try to use `complete` as a method name, nor do I think they'll be unhappy if they try to - `finish` is a word that means the same thing and is shorter, so they'll probably reach for that first.  `finalize` is also available, so there's no shortage of method names that could be used.  They also will likely have learned about how to initialize their classes and records before trying to write methods, so they should already know about `this.complete()` and its use and will associate that with its use in the initializer rather than trying to use it for their own purposes.

We made this change 4 years ago and we haven't received a single complaint about `this.complete()` that I recall.  I think at most an argument could be made for introducing a warning if a user does try to write a method named `complete`, but I think if we have that warning, a user will say "oh, of course, how silly of me" and use `finish` or `finalize` instead without being unhappy.

**Comment by bradcray** [comment]
> We made this change 4 years ago and we haven't received a single complaint about this.complete() that I recall.

I'm not sure that a lack of complaints here implies that it's a feature that's been heavily used and appreciated as it stands.  The reason I didn't just dismiss this issue out-of-hand as having already been solved is that it's part of the language that continues to seem a bit finicky/special/weird (in the sense that I'd just as soon not have to tell someone about it if I didn't have a reason to do so, because it feels just weird enough that I'd blush slightly when doing so).  So I think the question posed in the OP (which I read as being "how special should we commit to this as being, or should we make it less special somehow?") makes sense as we come into 2.0.  And the relationship to special methods resonates with me (in the sense that this is something that looks like a method call and is definitely special).

> They also will likely have learned about how to initialize their classes and records before trying to write methods, so they should already know about this.complete() and its use and will associate that with its use in the initializer rather than trying to use it for their own purposes.

I'm not sure I agree with this since the compiler-introduced initializers do the job pretty well for many uses of records and classes.  Even when writing custom initializers, I often don't need to use `this.complete()` for many common cases.

> I think at most an argument could be made for introducing a warning if a user does try to write a method named complete, but I think if we have that warning, a user will say "oh, of course, how silly of me" and use finish or finalize instead without being unhappy.

If we were to restrict users from writing methods named `complete()`, that feels to me like good evidence that the identifier should be reserved or special in some way.

If we wanted to keep it unreserved and add a warning for cases like this, I think it should either be generated only:

* on a call to `this.complete()` when there's a user-defined `this.complete()` that it could resolve to
* or potentially when defining a 0-argument-accepting method named `complete()` (but not a multi-argument version)

**Comment by dlongnecke-cray** [comment]
> I don't think we should change it. I don't actually think a user is going to try to use complete as a method name, nor do I think they'll be unhappy if they try to

This sounds like you're more-or-less in favor of option 1 (refine the status quo)?

Note that today it's not even possible to call `this.complete()` twice, as the compiler emits a rather esoteric:

```
error: use of this.complete() call in phase 2
```

So that precludes the possibility of users calling `complete()` in an initializer at all.

**Comment by e-kayrakli** [comment]
Without putting much thought into this, I've always wanted to make up a keyword to replace `this.complete()`. However, reading past discussions and thinking more about it, the fact that it looks like a method makes way more sense to me.

If we take approach 3 (keyword, that is), we'll either:
- use single word: but that would look too much like a control flow statement (eg, `return`, `break`, `continue`) rather something is being done on the object or the type
- use two words to make it clear that it is doing something on the object (eg, `delete foo`): but then, `this` would be the only reasonable "argument" to it. Then, we'll have a two-token pattern, which also is a bit too different for me to like :)

Based on that, I am currently in favor of 2a-based strategy that Brad outlined [above](https://github.com/chapel-lang/chapel/issues/21335#issuecomment-1374327362).

Not sure if `this.complete()` is the marker as a whole today, but if we make it just another special method, the programmer should be able to call both `this.~complete~()` and `~complete~()` in their initializers. Which would make it less of an oddity.

**Comment by mppf** [comment]
It's my understanding that, in an initializer, you can write `complete()` or `this.complete()`. I think some of us like to write `this.complete()` for clarity.

My own 2¢ about this: I think we should grandfather in certain special methods. I think of `init`, `init=`, `deinit`, and `this.complete()` as the ones I don't think we should change. I don't have a lot of preference about how the spec is updated and whether or not we want to allow `this.complete()` to call a user method within an initializer. (I would lean towards "no, but you can call a user method called complete in other contexts").

**Comment by bradcray** [comment]
Continuing my [previous brainstorming list](https://github.com/chapel-lang/chapel/issues/21335#issuecomment-1374317645) of ways to express this, based on today's discussion pointing out that reserving type method names on built-in types/identifiers is not so bad:

```chapel
class.complete();
record.complete();
type.complete();
t.complete();  // where we're defining the `init()` for `record t` or `class t`
```

**Comment by e-kayrakli** [comment]
> ```chapel
> class.complete();
> record.complete();
> type.complete();
> t.complete();  // where we're defining the `init()` for `record t` or `class t`
> ```

`t.complete()` can reduce maintainability. If I want to change my type name, I would have to go change these, too. Same argument can be made for `class` and `record`, but probably not as strongly.

`type.complete()` is interesting. So, we can do `this.type` in a method, but we can also drop `this.` in methods. Rhetorically, does this imply that you can use `type` anywhere in the methods instead of `this.type`? It is probably impossible to get the parser right for that.

This is not to say that I am against any of these, but not seeing them as improvements over `this.complete()` myself.

**Comment by dlongnecke-cray** [comment]
I like `type.complete()` because then I don't have to change it if my type name changes or I change it from a record to a class. What is the benefit of `class.complete()` and `record.complete()`? Isn't it possible for the user to define `record.complete()` if they wanted to have a `complete()` for every record?

I've never heard of `proc type.complete()`, though. Can we do that today to define a method on every type?

**Comment by mppf** [comment]
What is the problem we are trying to solve here? Is it that `this.complete()` looks weird / has a weird concept? Or is it

> Users may still want to define a method called complete()

You can do that today

``` chapel
record r {
  proc complete() {
    writeln("in r.complete()");
  }
}

proc main() {
  var x = new r();
  x.complete();
}
```

outputs `in r.complete()`

It is just not called in the case of appearing within an `init` method:

``` chapel
record r {
  proc complete() {
    writeln("in r.complete()");
  }
  proc init() {
    this.complete();
  }
}

proc main() {
  var x = new r();
}
```

(produces no output)

This behavior seems fine to me, personally. People writing database transactions inside of initializers can use a wrapper method. (Or not write it that way).

Maybe people are seeing some other problem that I'm not?

**Comment by bradcray** [comment]
Some users have complained that `this.complete();` is a magical feature (in the sense that it's special, has significant impacts on initializer behavior, is something the compiler will suggest you add if you start trying to do things too soon in an initializer) but looks like a completely conventional call, so maybe deserves more specialized syntax in some way.  That said, they also nodded to the fact that maybe that ship had already sailed.

**Comment by stonea** [comment]
> What is the problem we are trying to solve here? Is it that `this.complete()` looks weird / has a weird concept? Or is it
> 
> > Users may still want to define a method called complete()
> 
> You can do that today
> 
> ``` chapel
> record r {
>   proc complete() {
>     writeln("in r.complete()");
>   }
> }
> 
> proc main() {
>   var x = new r();
>   x.complete();
> }
> ```
> 
> outputs `in r.complete()`
> 
> It is just not called in the case of appearing within an `init` method:
> 
> ``` chapel
> record r {
>   proc complete() {
>     writeln("in r.complete()");
>   }
>   proc init() {
>     this.complete();
>   }
> }
> 
> proc main() {
>   var x = new r();
> }
> ```
> 
> (produces no output)

I think this behavior is preferable to the alternative of having 'complete()', as a method, always be called but it is rather surprising.

Someone ignorant about (or who just forgot about) the init-method specific behavior of `complete`, who did want to use it as a normal method called by the constructor would be puzzled/frustrated when they write their "call" to it but it never actually gets called.

**Comment by mppf** [comment]
> Someone ignorant about (or who just forgot about) the init-method specific behavior of `complete`, who did want to use it as a normal method called by the constructor would be puzzled/frustrated when they write their "call" to it but it never actually gets called.

IMO that specific issue is easily addressed with a warning along the lines of "I see you have defined a `yourtype.complete` method but this call here in an initializer will not call it".

**Comment by bradcray** [comment]
In a discussion with users last week, nervousness was expressed that `this.complete()` did not look more special/unique/different given how "magical" it is (their adjective, not mine).

**Comment by bradcray** [comment]
In recent discussions, Michael and I proposed replacing `this.complete();` with simply `complete;` and it satisfied the users mentioned in the previous comment, but wasn't a slam dunk for the Chapel developers team, who voted it into subcommittee.

**Comment by bradcray** [comment]
> In a language where foo; could not evaluate to proc foo {} I would agree that complete; feels sufficiently different (even then - you could have a field named complete), but in Chapel I feel like complete; has all the same problems as complete();.

This doesn't really bother me because I think we'd consider it poor style to define a paren-less method with side effects that would be worth calling like `foo;` or to refer to a field in a no-op manner like that.  I.e., I don't think it's common / likely for people to write `foo;` in their programs (the main case I can think of currently is "forcing default initialization in a potential split init context", but that's one where we've been discussing alternatives as well, to potentially make `foo;` an illegal syntactic construct.  Not mentioned in my short summary above (but mentioned in the meeting where we discussed it recently), we also discussed generating warnings when users wrote code that could conceivably be invoked as `complete;` in an initializer context (like in the presence of a paren-less method, field, or in-scope var named `complete`).
